pkg/controller/util.go

```diff
@@ -175,28 +171,79 @@ func GetVolumeCapabilities(pvSpec *v1.PersistentVolumeSpec) (*csi.VolumeCapabili
 		}
 	}
 
-	// Translate array of modes into single VolumeCapability
+	var am csi.VolumeCapability_AccessMode_Mode
+	var err error
+	if singleNodeMultiWriterCapable {
+		am, err = getSingleNodeMultiWriterCapableAccessMode(pvSpec)
+	} else {
+		am, err = getAccessMode(pvSpec)
+	}
+	if err != nil {
+		return nil, err
+	}
+	cap.AccessMode.Mode = am
+	return cap, nil
+}
+
+// getAccessMode maps PV access modes to CSI access modes
+func getAccessMode(pvSpec *v1.PersistentVolumeSpec) (csi.VolumeCapability_AccessMode_Mode, error) {
+	m := uniqueAccessModes(pvSpec)
+
 	switch {
 	case m[v1.ReadWriteMany]:
 		// ReadWriteMany trumps everything, regardless what other modes are set
-		cap.AccessMode.Mode = csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER
+		return csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER, nil
 
 	case m[v1.ReadOnlyMany] && m[v1.ReadWriteOnce]:
-		// This is no way how to translate this to CSI...
-		return nil, fmt.Errorf("CSI does not support ReadOnlyMany and ReadWriteOnce on the same PersistentVolume")
+		// This is not possible in the CSI spec
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("CSI does not support ReadOnlyMany and ReadWriteOnce on the same PersistentVolume")
 
 	case m[v1.ReadOnlyMany]:
 		// There is only ReadOnlyMany set
-		cap.AccessMode.Mode = csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY
+		return csi.VolumeCapability_AccessMode_MULTI_NODE_READER_ONLY, nil
 
 	case m[v1.ReadWriteOnce]:
 		// There is only ReadWriteOnce set
-		cap.AccessMode.Mode = csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER
+		return csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER, nil
 
 	default:
-		return nil, fmt.Errorf("unsupported AccessMode combination: %+v", pvSpec.AccessModes)
+		return csi.VolumeCapability_AccessMode_UNKNOWN, fmt.Errorf("unsupported AccessMode combination: %+v", pvSpec.AccessModes)
```

Comment:
Is it OK to return error when the PV supports only `ReadWriteOncePod`? Should it be handled the same as `ReadWriteOnce`?

Reply:
I don't think users should use ReadWriteOncePod if the driver does not support the capability.

Reply:
Since most of `ReadWriteOncePod` enhancement is done in the scheduler and kubelet to make sure a volume is not presented to multiple pods on the same node, IMO it makes sense to fall back to `ReadWriteOnce` behavior during attach and maybe in kubelet too. The new access mode will be hard to use, because only some CSI drivers will support it, at least initially. Application writers cannot write e.g. a helm chart that reliably deploys their apps on a generic cluster with `ReadWriteOncePod`, because they don't know what CSI drivers are deployed on a random cluster.